presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java

```diff
@@ -697,21 +698,30 @@ protected ExtractionResult visitInPredicate(InPredicate node, Boolean complement
             InListExpression valueList = (InListExpression) node.getValueList();
             checkState(!valueList.getValues().isEmpty(), "InListExpression should never be empty");
 
-            ImmutableList.Builder<Expression> disjuncts = ImmutableList.builder();
-            for (Expression expression : valueList.getValues()) {
-                disjuncts.add(new ComparisonExpression(EQUAL, node.getValue(), expression));
+            TupleDomain<Symbol> tupleDomain;
+            BiFunction<TupleDomain<Symbol>, TupleDomain<Symbol>, TupleDomain<Symbol>> mergingFunction;
+            if (complement) {
+                tupleDomain = TupleDomain.all();
+                mergingFunction = TupleDomain::intersect;
+            }
+            else {
+                tupleDomain = TupleDomain.none();
+                mergingFunction = TupleDomain::columnWiseUnion;
             }
-            ExtractionResult extractionResult = process(or(disjuncts.build()), complement);
 
-            // preserve original IN predicate as remaining predicate
-            if (extractionResult.tupleDomain.isAll()) {
-                Expression originalPredicate = node;
-                if (complement) {
-                    originalPredicate = new NotExpression(originalPredicate);
+            for (Expression expression : valueList.getValues()) {
+                ExtractionResult result = process(new ComparisonExpression(EQUAL, node.getValue(), expression), complement);
+                tupleDomain = mergingFunction.apply(tupleDomain, result.getTupleDomain());
+                if (tupleDomain.isAll() || !result.getRemainingExpression().equals(TRUE_LITERAL)) {
```

Comment:
Is `tupleDomain.isAll()` this needed here? Should be the condition be the same for complement and not complement case here?
Also it would be nice to `!result.getRemainingExpression().equals(TRUE_LITERAL)` wrap as `!result.isFullyRewritten()` or `result.hasRemainingExpression()`